### PR TITLE
Last field might be a backlink field.

### LINKED ISSUE #4864: Realm 3.4.0: isEmpty for linked relationships causes SIGSEGV
#### Goal

Query for empty inverse relationships: get all dogs without owners.

#### Expected Results

Query should return the dogs without owners.

#### Actual Results

SIGSEGV:
```
06-27 22:21:19.670 3461-3461/com.realm.bug.realmbug A/libc: Fatal signal 11 (SIGSEGV), code 2, fault addr 0x2429a227 in tid 3461 (lm.bug.realmbug)
                                                            
                                                            [ 06-27 22:21:19.670  1256: 1256 W/         ]
                                                            debuggerd: handling request: pid=3461 uid=10079 gid=10079 tid=3461
```
Full log attached: [full_log.txt](https://github.com/realm/realm-java/files/1106998/full_log.txt)

#### Steps & Code to Reproduce

The first query (`equalTo("owner.id", "person")`) is successful, and works as expected:
`06-27 21:52:47.999 31101-31101/com.realm.bug.realmbug I/MainActivity: size=1`
However the second (`isEmpty("owner")`) causes segfault.

#### Code Sample

`Dog` and `Person` are simple `RealmObject`s with `@PrimaryKey` id.

```java
private void testRealmBug(Context context) {
    Realm.init(context);
    RealmConfiguration realmConfiguration = new RealmConfiguration.Builder()
            .inMemory()
            .schemaVersion(2)
            .build();
    Realm realm = Realm.getInstance(realmConfiguration);

    realm.executeTransaction(new Realm.Transaction() {
        @Override
        public void execute(Realm realm) {
            Dog dog = realm.createObject(Dog.class, "dog");
            Person person = realm.createObject(Person.class, "person");
            person.getDogs().add(dog);
        }
    });

    // works as expected
    RealmResults<Dog> dogs = realm.where(Dog.class)
            .equalTo("owner.id", "person")
            .findAll();
    Log.i(TAG, "size=" + dogs.size());

    // SIGSEGV
    dogs = realm.where(Dog.class)
            .isEmpty("owner")
            .findAll();
    Log.i(TAG, "size=" + dogs.size());
}
```

```java
@RealmClass
public class Person extends RealmObject {

    @PrimaryKey
    private String id;

    private RealmList<Dog> dogs;

    public Person() {
    }

    public RealmList<Dog> getDogs() {
        return dogs;
    }

    public String getId() {
        return id;
    }
}
```

```java
@RealmClass
public class Dog extends RealmObject {

    @PrimaryKey
    private String id;

    @LinkingObjects("dogs")
    private final RealmResults<Person> owner = null;

    public Dog() {
    }

    public String getId() {
        return id;
    }
}
```

#### Version of Realm and tooling
Realm version(s): 3.4.0

Realm sync feature enabled: no

Android Studio version: 2.3.3

Which Android version and device: API24 Nexus 5X (emulator), API25 Pixel (emulator)

### PR DESCRIPTION
**Pull request by kneth**
When querying a `LinkingObjects` field (aka backlink field), it might be the last. `isEmpty()` is a pretty useful predicate, and we need to track down the source table.

The changes to the C++ code is only for code style consistency.

Closes #4864

### CODE REVIEW

File: realm/realm-library/src/main/java/io/realm/internal/fields/CachedFieldDescriptor.java [hunk 1]
```diff
@@ -82,8 +82,8 @@ protected void compileFieldDescription(List<String> fields) {
             columnType = tableInfo.getColumnType(columnName);
             if (i < nFields - 1) {
                 verifyInternalColumnType(currentTable, columnName, columnType);
-                currentTable = tableInfo.getLinkedTable(columnName);
             }
+            currentTable = tableInfo.getLinkedTable(columnName);
```
Comment (bmeike) [inline realm/realm-library/src/main/java/io/realm/internal/fields/CachedFieldDescriptor.java]: Does this work?  I absolutely don't remember but that was in the conditional for a reason, at one time.  What happens if you call `schema.getNativeTablePtr(null)` a couple lines down?

Comment (kneth) [inline realm/realm-library/src/main/java/io/realm/internal/fields/CachedFieldDescriptor.java]: `schema.getNativeTablePtr()` is only called when the type is `LINKING_OBJECTS`, and in that case, `currentTable` cannot be `null`.

File: realm/realm-library/src/main/java/io/realm/internal/fields/CachedFieldDescriptor.java [hunk 2]
```diff
@@ -80,6 +80,7 @@ protected void compileFieldDescription(List<String> fields) {
             }
 
             columnType = tableInfo.getColumnType(columnName);
+            // we don't check the type of the last field in the chain since it done in the C++ code
```
Comment (Zhuinden) [inline realm/realm-library/src/main/java/io/realm/internal/fields/CachedFieldDescriptor.java]: +is

File: realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp [hunk 3]
```diff
@@ -1668,9 +1668,10 @@ JNIEXPORT void JNICALL Java_io_realm_internal_TableQuery_nativeIsEmpty(JNIEnv* e
         auto column_idx = static_cast<size_t>(index_arr[arr_len - 1]);
 
         // Support a backlink as the last column in a field descriptor
-        Table* last = TBL(table_arr[arr_len-1]);
+        auto last = reinterpret_cast<Table*>(table_arr[arr_len-1]);
         if (last != nullptr) {
-            pQuery->and_query(src_table_ref->column<BackLink>(*last, column_idx).count() == 0);
+            TableRef last_ref = TableRef(last);
```
Comment (beeender) [inline realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp]: `TableRef` is not needed here, i like less lines of code :)

File: realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp [hunk 4]
```diff
@@ -1739,9 +1740,10 @@ Java_io_realm_internal_TableQuery_nativeIsNotEmpty(JNIEnv *env, jobject, jlong n
         auto column_idx = static_cast<size_t>(index_arr[arr_len - 1]);
 
         // Support a backlink as the last column in a field descriptor
-        auto last = reinterpret_cast<Table *>(table_arr[arr_len-1]);
+        auto last = reinterpret_cast<Table*>(table_arr[arr_len-1]);
         if (last != nullptr) {
-            pQuery->and_query(src_table_ref->column<BackLink>(*last, column_idx).count() != 0);
+            TableRef last_ref = TableRef(last);
```
Comment (beeender) [inline realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp]: same

File: realm/realm-library/src/main/java/io/realm/internal/fields/CachedFieldDescriptor.java (hunk 1 above)
Comment (beeender) [inline realm/realm-library/src/main/java/io/realm/internal/fields/CachedFieldDescriptor.java]: Will, for example, `query.equalTo("abc", "backlinkField")` still throw? I guess so?

File: realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp (hunk 4 above)
Comment (bmeike) [inline realm/realm-library/src/main/cpp/io_realm_internal_TableQuery.cpp]: I think this is a matter of taste.  I had an off-line discussion with @kneth about this and I agree with him that the explicit cast is better than my original code, which just used `last`, a `Table*`.  Being succinct is good, for sure.  I don't think this cast costs much of anything and, for me at least, it makes it clearer what's going on.   I'm good with either but, FWIW, I prefer this.

**Comment by kneth** [comment]
@beeender Yeah, `equalTo()` will throw `IllegalArgumentException`. I add a test.

**Pull request merged**